### PR TITLE
#3355 Redact configs in sgcollect_info

### LINKED ISSUE #3259: Log Redaction
Tracking issue for Log Redaction epic

### LINKED ISSUE #3355: Log redaction: sgcollect_info should redact data in running_server_config.log
Running `sgcollect_info` will save the running server's config to `running_server_config.log`

The data in here will need redacting, as it could contain usernames, database names, URLs, etc.

Related: #3296 #3259

### PR DESCRIPTION
**Pull request by bbrks**
Fixes #3355 - Tags sensitive user data in configs collected by sgcollect_info before being redacted.

Closes #3259 

## Tagged fields:
```
databases.foo_db.username

databases.foo_db.users.foo_user
databases.foo_db.users.foo_user.name
databases.foo_db.users.foo_user.admin_channels.[]
databases.foo_db.users.foo_user.admin_roles.[]

databases.foo_db.roles.foo_role
databases.foo_db.roles.foo_role.admin_channels.[]
```

Using [this config reference](https://developer.couchbase.com/documentation/mobile/current/guides/sync-gateway/config-properties/index.html#configuration-reference)

### CODE REVIEW

File: tools/password_remover.py [hunk 1]
```diff
@@ -23,6 +24,112 @@ def is_valid_json(invalid_json):
...
+                    users[user]["name"] = UD(users[user]["name"])
+                if "admin_channels" in users[user]:
+                    admin_channels = users[user]["admin_channels"]
+                    for i, _ in enumerate(admin_channels):
+                        admin_channels[i] = UD(admin_channels[i])
+                if "admin_roles" in users[user]:
+                    admin_roles = users[user]["admin_roles"]
+                    for i, _ in enumerate(admin_roles):
+                        admin_roles[i] = UD(admin_roles[i])
+            for k, _ in users.items():
```
Comment (tleyden) [inline tools/password_remover.py]: What about just re-using `i` here?  I don't think it's in scope anywhere.

Or maybe use `j` .. a little more obvious it's an index.

File: tools/password_remover.py [hunk 2]
```diff
@@ -23,6 +24,112 @@ def is_valid_json(invalid_json):
...
+        - Tag any sensitive user-data fields with <ud></ud> tags.
+        """
+
+        if "username" in db:
+            db["username"] = UD(db["username"])
+
+        if "users" in db:
+            users = db["users"]
+            for user in users:
+                if "name" in users[user]:
```
Comment (tleyden) [inline tools/password_remover.py]: Would this be a bit clearer?

```
for username in users:
    user = users[username]
    user["name"] = UD(user["name"])
```

File: tools/password_remover.py [hunk 3]
```diff
@@ -23,6 +24,112 @@ def is_valid_json(invalid_json):
...
+
+        tag_userdata_in_server_json(parsed_json)
+
+        formatted_json_string = json.dumps(parsed_json, indent=4)
+        return formatted_json_string
+
+    except Exception as e:
+        msg = "Exception trying to tag config user data in {0}.  Exception: {1}".format(json_text, e)
+        if log_json_parsing_exceptions:
+            print(msg)
```
Comment (tleyden) [inline tools/password_remover.py]: Since `msg` is only used once, might be slightly clearer if it's just `print("Exeption....")`

**Review by tleyden — DISMISSED** [review]
Inline comments + how about adding a unit test along the lines of https://github.com/couchbase/sync_gateway/blob/feature/issue_3355_redact_config_sgcollect/tools/password_remover.py#L300 that invokes `tag_userdata_in_server_config` and does a smoke test sanity check to make sure things were redacted as expected?  For example, it could just run a regex for "<ud>" and make sure it was present.

**Comment by bbrks** [comment]
@tleyden Addressed your comments and added unit test which covers some positive, and some negative assertions. Thanks!

```
10:55 $ python2.7 password_remover.py -v
test_alternative_config (__main__.TestRemovePasswords) ... ok
test_basic (__main__.TestRemovePasswords) ... ok
test_config_fragment (__main__.TestRemovePasswords) ... ok
test_non_parseable_config (__main__.TestRemovePasswords) ... ok
test_basic (__main__.TestTagUserData) ... ok

----------------------------------------------------------------------
Ran 5 tests in 0.004s

OK
```

**Review by tleyden — APPROVED** [review]
LGTM

**Pull request merged**